File: GRAPH/547_Number_of_Provinces.cpp
// <--------------------------------------------------------------------DFS--------------------------------------------------------------------------->
/*
TC: O(V+2E)
SC: O(2E)
*/

class Solution {
public:
    
    void dfs(int node, vector<int>&vis, vector<vector<int>>&adj){
            vis[node]=1;

            for(int child: adj[node]){
                if(!vis[child]){
                    dfs(child,vis,adj);
                }
            }
        return;
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<vector<int>> adj(n+1, vector<int> ());

        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(isConnected[i][j]==1){
                    adj[i+1].push_back(j+1);
                }
            }
        }

        vector<int> vis;
        vis.resize(n+1,0);
        int count = 0;
        for(int i = 1; i<=n; i++){
            if(!vis[i]){
                count++;
                dfs(i,vis,adj);
            }
        }
        return count;
    }
};



// <-------------------------------------------------------------BFS--------------------------------------------------------------------------------------->

/*
BFS 
TC: O(V+2E)
SC: O(2E)
  */


class Solution {
public:
    
    void bfs(int node, vector<int>&vis, vector<vector<int>>&adj, queue<int>&q){

            vis[node]=1;
            q.push(node);
            while(!q.empty()){

                int node = q.front();
                q.pop();

                for(int child: adj[node]){
                if(!vis[child]){
                    q.push(child);
                    vis[child]=1;
                }
            }
        } 
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<vector<int>> adj(n+1, vector<int> ());

        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(isConnected[i][j]==1){
                    adj[i+1].push_back(j+1);
                }
            }
        }

        queue<int> q;
        vector<int> vis;
        vis.resize(n+1,0);
        int count = 0;
        for(int i = 1; i<=n; i++){
            if(!vis[i]){
                count++;
                bfs(i,vis,adj,q);
            }
        }
        return count;
    }
};

File: GRAPH/547_Number_of_Provinces.java

// <----------------------------------------------------------BFS----------------------------------------------------------------------------------------->

class Solution {

     void bfs(int node, ArrayList<Integer> vis, ArrayList<ArrayList<Integer>> adj){

            Queue<Integer> q = new LinkedList<Integer> ();
            vis.set(node,1);
            q.add(node);
            while(!q.isEmpty()){

                int parent = q.poll();
            
                for(int child: adj.get(parent)){
                if(vis.get(child)==0){
                    q.add(child);
                    vis.set(child,1);
                }
            }
        } 
    }

    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>> ();

        for(int i = 0;i <=n; i++){
            adj.add(new ArrayList<Integer> ());
        }

        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(isConnected[i][j]==1){
                    adj.get(i+1).add(j+1);
                }
            }
        }

        ArrayList<Integer> vis = new ArrayList<Integer> (Collections.nCopies(n+1,0));
        
        int count = 0;
        for(int i = 1; i<=n; i++){
            if(vis.get(i)==0){
                count++;
                bfs(i,vis,adj);
            }
        }
        return count;
    }
}

    
   
// <-----------------------------------------------------------------DFS----------------------------------------------------------------------------------->


class Solution {

     void dfs(int node, ArrayList<Integer> vis, ArrayList<ArrayList<Integer>> adj){

            vis.set(node,1);
            
                for(int child: adj.get(node)){
                if(vis.get(child)==0){
                    dfs(child,vis,adj);                    
                }
            }
    }

    public int findCircleNum(int[][] isConnected) {

        int n = isConnected.length;
        ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>> ();

        for(int i = 0;i <=n; i++){
            adj.add(new ArrayList<Integer> ());
        }

        for(int i = 0; i<n; i++){
            for(int j = 0; j<n; j++){
                if(isConnected[i][j]==1){
                    adj.get(i+1).add(j+1);
                }
            }
        }

        ArrayList<Integer> vis = new ArrayList<Integer> (Collections.nCopies(n+1,0));
        
        int count = 0;
        for(int i = 1; i<=n; i++){
            if(vis.get(i)==0){
                count++;
                dfs(i,vis,adj);
            }
        }
        return count;
    }
}

    
   



